feat: new smart default features (#5572)
JetBrains/kotlin-web-site: docs/topics/maven/maven-configure-project.md

To apply the Kotlin Maven plugin with `<extensions>`, update your `pom.xml` build file as follows:

> 1. In the `<properties>` section, define the target versions of Kotlin and JVM:

Reviewer: ```suggestion
1. In the `<properties>` section, define the target versions of Kotlin and the JVM:
```

* Registers `src/main/kotlin` and `src/test/kotlin` directories as source roots if they already exist but are not specified in the plugin configuration.
* Adds the [`kotlin-stdlib` dependency](maven-set-dependencies.md#dependency-on-the-standard-library) if it's not already defined in the project.
* Adds `compile`, `test-compile`, `kapt`, and `test-kapt` executions to your build, bound to their appropriate [lifecycle phases](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html).
  So you don't need to manually set up the `<executions>` section with `<id>` and `<goals>` for `kapt`, Kotlin's `compile`,

>   and Java's `compile` executions to run in the correct order.

Reviewer: ```suggestion
  and Java's `compile` executions to ensure they run in the correct order.
```
I think?

>
{style="note"}

### JVM target version

The `<extensions>` option ensures that the Kotlin and Maven compilers target the same bytecode version.

The Kotlin Maven plugin automatically resolves the JVM target version in the following order:

```mermaid
graph TD
    A["<b>Priority 1</b><br/>kotlin.compiler.jdkRelease</br>or kotlin.compiler.jvmTarget"]
    B["<b>Priority 2</b><br/>maven.compiler.release"]
    C["<b>Priority 3</b><br/>maven.compiler.target"]

    A --> B
    B --> C
```

#### Kotlin compiler versions

The version set in `kotlin.compiler.jdkRelease` or `kotlin.compiler.jvmTarget` property takes priority if either
is defined in the project.

> Keep in mind that these options of the Kotlin compiler behave differently:

Reviewer: ```suggestion
Keep in mind that these Kotlin compiler options behave differently:
```

| Kotlin compiler option       | Controls bytecode version of the output | Limits API to specified JDK                                                                   |
|------------------------------|-----------------------------------------|-----------------------------------------------------------------------------------------------|
| `kotlin.compiler.jvmTarget`  | Yes                                     | No restrictions on JDK APIs in your code                                                      |
| `kotlin.compiler.jdkRelease` | Yes                                     | Yes − only specific API version is allowed (equivalent to Java's `--release` compiler option) |

> You cannot set different JDK options for `kotlin.compiler.jdkRelease` and `kotlin.compiler.jvmTarget` at the same time.

> > This results in an error.

Reviewer: ```suggestion
> If you set different JDK options for `kotlin.compiler.jdkRelease` and `kotlin.compiler.jvmTarget` at the same time,
> you will see an error.
```

>
{style="note"}

#### Maven compiler versions

* If neither `kotlin.compiler.jdkRelease` nor `kotlin.compiler.jvmTarget` is set, the plugin takes

>   the `maven.compiler.release` version.

Reviewer: ```suggestion
* If neither the `kotlin.compiler.jdkRelease` nor the `kotlin.compiler.jvmTarget` option is set, the plugin takes
  the `maven.compiler.release` version.
```

>   It can be defined either as a project property or within the `maven-compiler-plugin` configuration.

Reviewer: ```suggestion
  The `maven.compiler.release` version can be defined either as a project property or within the `maven-compiler-plugin` configuration.
```
Just to be extra clear.

* If the Maven release version isn't set, the plugin takes the `maven.compiler.target` version.

  It can be defined either as a project property or within the `maven-compiler-plugin` configuration.

Keep in mind that `target` and `release` options of the Maven compiler behave differently:

| Maven compiler option    | Sets Kotlin's `jvmTarget` | Sets Kotlin's `jdkRelease` | Limits API to specified JDK              |
|--------------------------|---------------------------|----------------------------|------------------------------------------|

> | `maven.compiler.target`  | Yes                       | No                         | No − build's JDK classpath stays visible |

Reviewer: ```suggestion
| `maven.compiler.target`  | Yes                       | No                         | No − the build's JDK classpath stays visible |
```

> | `maven.compiler.release` | Yes                       | Yes                        | Yes − specific API version only          |

Reviewer: ```suggestion
| `maven.compiler.release` | Yes                       | Yes                        | Yes − to the specific API version only          |
```

> The `<extensions>` option only checks project-level properties and the global `maven-compiler-plugin` configuration.

> > It doesn’t check the configurations defined in the plugin’s `<executions>` section.

Reviewer: ```suggestion
> It doesn't check the configurations defined in the plugin's `<executions>` section.
```
